rfcs/0000-expressive-diagnostics.md

```diff
@@ -0,0 +1,356 @@
+- _Help_: Additional friendly information, which is not related to a place in code, that helps understanding and solving the issue.
+
+![oopsies image not showing](../files/0010-expressive-diagnostics-cli.png)
+
+When using IDEs like VS Code Sway programmers will have the experience similar to one offered by the [Rust analyzer](https://rust-analyzer.github.io/).
+
+Popup in the editor will provide the _reason_, _issue_, _hints_, as well as the _help_, similar to this Rust example:
+
+![oopsies image not showing](../files/0010-expressive-diagnostics-vscode-hints.png)
+
+Programmer will have the option to open the full compiler diagnostic, getting the same output like when using `forc` from the command line, similar to this Rust example:
+
+![oopsies image not showing](../files/0010-expressive-diagnostics-vscode-full-diagnostic.png)
+
+## Wording guidelines
+
+To ensure consistency and apply best practices<sup>[5]</sup> the diagnostics will follow these guidelines:
+
+- _Reason_ will be short and point out the language constraint which was violated. It will _not_ finish in fullstop, to emphasize brevity.
+- _Issue_ will be short and point out the specific situation encountered in code. It will also _not_ finish in fullstop, to emphasize brevity.
+- CLI or IDE plugin might choose to combine the _reason_ and the _issue_ as shown at the image above.
+- _Reason_ and _issue_ are given in plain english language, free of, e.g., compiler jargon.
+- _Hints_ and _help_ are as well written in plain english, using proper punctuation and grammar rules.
+- _Hints_ and _help_ try to give as much of useful context as possible and to be as specific as possible.
+- Identifier and type names in messages are enclosed in "double quotes".
+- Articles "the" and "a/an" are not used at the begining of a sentence. E.g., "Variable "X" shadows..." instead of "The variable "X" shadows...". They can be used in formulations like "This is the original declaration...".
+
+Here is an example. The existing error message:
+
+```
+Name "MyName" is defined multiple times.
+```
+
+could, depending on the context, have the following form:
+
+```
+Reason: Item names must be unique
+Issue:  There is already an imported struct with the name "MyName"
+Hints:  [error] This is the enum "MyName" that has the same name as the imported struct.
+        [info]  The struct "MyName" gets imported here.
+        [info]  This is the original declaration of the imported struct "MyName".
+Help:   Items like structs, enums, traits, and ABIs must have a unique name in scope.
+        Consider renaming the enum "MyName" or using an alias for the imported struct.
+```
+
+## Diagnostic codes
+
+A diagnostic _code_ uniquely identifiy a particular _reason_. Considering that we already have these diagnostic areas:
```

Comment:
Can the same reason be used as both an error and a warning? Like in some contexts it's only a warning, but in other contexts it's actually a blocker to compilation? 

For instance, if the compiler ever had a "strict" mode, where some warnings would be treated as errors.

Reply:
Great point! Treat warnings as errors is my preferred way of compiling :-)

The easiest way to get that functionality now, would be to solve it at the final receiver of the diagnostics, before the next step, whatever that is. Mean, compiler emits diagnostics and before we generate IR we treat warnings as errors.

`#[allow()]`s would be handled in the compiler before, at the emitting time, of course.

Other way around, treating errors as warnings, I guess we don't have a use case for that.

That means, yes, we can treat errors as warnings without changing their definitions, should be straightforward.